pennylane/_device.py

```diff
@@ -285,54 +285,91 @@ def supports_operation(self, operation):
         Returns:
             bool: ``True`` iff supplied operation is supported
         """
-        if isinstance(operation, type) and issubclass(operation, Operation):
-            return operation.__name__ in self.operations
-        if isinstance(operation, str):
-            return operation in self.operations
+        op_name = operation
 
-        raise ValueError("The given operation must either be a pennylane.Operation class or a string.")
+        if issubclass(operation.__class__, Operation):
+            op_name = operation.name
+        elif isinstance(operation, type):
+            op_name = operation.__name__
+        elif not isinstance(op_name, str):
+            raise ValueError("The given operation must either be a pennylane.Operation class or a string.")
+
+        if op_name not in self.operations:
+            raise DeviceError("Gate {} not supported on device {}".format(op_name, self.short_name))
+
+        return True
 
     def supports_observable(self, observable):
         """Checks if an observable is supported by this device.
 
         Args:
-            operation (Observable,str): observable to be checked
+            observable (Observable,str): observable to be checked
 
         Returns:
             bool: ``True`` iff supplied observable is supported
         """
-        if isinstance(observable, type) and issubclass(observable, Observable):
-            return observable.__name__ in self.observables
-        if isinstance(observable, str):
-            return observable in self.observables
+        obs_name = observable
+
+        if issubclass(observable.__class__, Observable):
+            obs_name = observable.name
+        elif isinstance(observable, type):
+            obs_name = observable.__name__
+        elif not isinstance(obs_name, str):
+            raise ValueError("The given observable must either be a pennylane.Observable class or a string.")
+
+        if obs_name not in self.observables:
+            raise DeviceError("Observable {} not supported on device {}".format(obs_name, self.short_name))
 
-        raise ValueError("The given operation must either be a pennylane.Observable class or a string.")
+        return True
```

Comment:
How come we switched this from 
"returns True/False (but might throw exception)"
to
"returns True (but might throw exception)"
?

Note that "(but might throw exception)" applies to any function, irrespective of what its docstring says about it's return value"

Reply:
I feel like this way of coding things makes the function less useful. 
With this change, if I want to check if an observable is supported from somewhere else in the code, I will have to use `try/except`. Otherwise, the fact that an observable isn't supported would cause my code to give an exception. 

In essence, with this proposed change, I feel it is less easy to use this function

Reply:
Note that there's a difference between *asking* if a gate is supported, and trying to *force* a gate to be executed that isn't supported. The former case shouldn't throw an exception unless the request is incorrectly formatted

Reply:
+1 on nathans comments

Reply:
The idea here was to refactor the ``check_validity`` method by using ``supports_operation`` and ``supports_observable``. When the original PR was made, these latter two were used for tests only (since then I have seen that they were also used in the main parts of the codebase as well).

I do agree with these points and have reverted the changes in a way that checks are still included for the inverse.

A minor refactor for ``check_validity`` is included using these methods, I feel that this method is now a bit more readable.